master/pkg/stream/declarative.go

```diff
@@ -0,0 +1,182 @@
+	// are we emitting from a range?
+	if ksi.start < ksi.end {
+		out := ksi.start
+		ksi.start++
+		return true, out, nil
+	}
+	// are we out of values?
+	if ksi.idx >= len(ksi.vals) {
+		return false, 0, nil
+	}
+	// parse the next value
+	val := ksi.vals[ksi.idx]
+	ksi.idx++
+	subs := strings.Split(val, "-")
+	if len(subs) == 1 {
+		// it's a single value
+		n, err := strconv.ParseInt(subs[0], 10, 64)
+		if err != nil {
+			return false, 0, fmt.Errorf("invalid number (%v): %s", val, err.Error())
+		}
+		return true, n, nil
+	} else if len(subs) == 2 {
+		// it's a range
+		start, err := strconv.ParseInt(subs[0], 10, 64)
+		if err != nil {
+			return false, 0, fmt.Errorf("invalid range (%v): %s", val, err.Error())
+		}
+		end, err := strconv.ParseInt(subs[1], 10, 64)
+		if err != nil {
+			return false, 0, fmt.Errorf("invalid range (%v): %s", val, err.Error())
+		}
+		if start >= end {
+			return false, 0, errors.Errorf("invalid range (%v)", val)
+		}
+		ksi.start = start + 1 // plus one because we emit start right now
+		ksi.end = end + 1     // plus one to save an exclusive endpoint
+		return true, start, nil
+	}
+	// any other number of splits around "-"
+	return false, 0, errors.Errorf("invalid value (%v)", val)
+}
+
+type keySetBuilder struct {
+	out     strings.Builder
+	started bool
+	comma   bool
+	start   int64
+	last    int64
+}
+
+func (ksb *keySetBuilder) emit() {
+	// do we need a separator?
+	if !ksb.comma {
+		ksb.comma = true
+	} else {
+		_, _ = ksb.out.WriteString(",")
+	}
+	// always emit "$START"
+	_, _ = ksb.out.WriteString(strconv.FormatInt(ksb.start, 10))
+	if ksb.last != ksb.start {
+		// for ranges, emit "-$PREV"
+		_, _ = ksb.out.WriteString("-")
+		_, _ = ksb.out.WriteString(strconv.FormatInt(ksb.last, 10))
+	}
+}
+
+func (ksb *keySetBuilder) Add(n int64) {
+	if !ksb.started {
+		// first call to Add()
+		ksb.started = true
+		ksb.start = n
+		ksb.last = n
+		return
+	}
+	if n == ksb.last+1 {
+		// extend the current range
+		ksb.last++
+		return
+	}
+	// emit previous number or range
+	ksb.emit()
+	// start new range
+	ksb.start = n
+	ksb.last = n
+}
+
+func (ksb *keySetBuilder) Finish() string {
+	if !ksb.started {
+		return ""
+	}
+	// emit final value
+	ksb.emit()
+	return ksb.out.String()
+}
+
+// ProcessKnown takes what the client reports as known keys, combined with which keys the server
+// knows exist, and returns what the client should be told is deleted and which keys the client is
+// not yet aware of, which should be hydrated by querying the database.
+//
+// Parameters:
+// - `known` is a range-encoded string, directly from the REST API.
+// - `exist` is a list of ints, likely from the database or some in-memory cache.
+//
+// Return Values:
+// - `gone` is a range-encoded string, suitable for returning over the REST API.
+// - `new` is a list of ints, which are PKs that will need hydrating from the database.
+func ProcessKnown(known string, exist []int64) (string, []int64, error) {
```

Comment:
`ProcessKnown` might not be entirely correct if `known` and/or `exist` is not sorted, while it is highly possible for `exist` to be unsorted

Reply:
I'll add in the assumption to the comment.

 `known` is created by the streaming client, which keeps track of the known keys and builds the known string, and `exist` is populate by a query for each streamable type (currently just projects). 
 
 The project select includes an `ORDER <PK> ASC`; so we should be be fine in terms of making sure  that `exists` is sorted.
 
 As long as the streaming client does not build an improper known string and any additional streamable types are also sorting the results of their queries properly, this shouldn't be an issue.

Reply:
Since `exist` is basically a concatenation of `oldEventsQuery` scan and `newEventsQuery` scan, both queries sorted does not guarantee `exist` also sorted

Reply:
Good catch!